### PR TITLE
Implement multiple progressions methods

### PR DESCRIPTION
**Pull request by kristijan-ujevic**

### CODE REVIEW

**Review by JohannesLoot — DISMISSED** [review]
Awesome job!
👍

File: Runtime/Game/LootLockerSDKManager.cs [hunk 1]
```diff
@@ -1496,6 +1641,159 @@ public static void GetEquipableContextToDefaultCharacter(Action<LootLockerContex
...
+        {
+            if (!CheckInitialized())
+            {
+                onComplete?.Invoke(LootLockerResponseFactory.SDKNotInitializedError<LootLockerPaginatedCharacterProgressions>());
+                return;
+            }
+
+            var endpoint = string.Format(LootLockerEndPoints.getAllCharacterProgressions.endPoint, characterId);
+
+            endpoint += "?";
```
Comment (kirre-bylund) [inline Runtime/Game/LootLockerSDKManager.cs:1663]: Same thing about trailing characters

**Review by kirre-bylund — CHANGES_REQUESTED** [review]
Hmm, this sparks an interesting discussion on whether stuff we add should adhere to the current double layered (but superfluous) structure or not.

File: Runtime/Client/LootLockerServerRequest.cs [hunk 2]
```diff
@@ -92,6 +92,12 @@ public static T Serialize<T>(LootLockerResponse serverResponse)
             return response;
         }
     }
+    public class LootLockerPagination
```
[the remark this answers is not in the document]
Comment (kristijan-ujevic) [reply Runtime/Client/LootLockerServerRequest.cs:95]: I noticed that but didn't want to touch something else in this PR to keep it as atomic as possible. I think we should use this one in both cases.

File: Runtime/Game/LootLockerSDKManager.cs [hunk 3]
```diff
@@ -1209,6 +1209,151 @@ public static void DeletePlayerFile(int fileId, Action<LootLockerResponse> onCom
         }
         #endregion
 
+        #region Player progressions
+
+        /// <summary>
+        /// Returns all progressions the player is currently on.
+        /// </summary>
+        /// <param name="count">Amount of entries to receive</param>
+        /// <param name="after">How many extra rows after the players position</param>
```
[the remark this answers is not in the document]
Comment (kristijan-ujevic) [reply Runtime/Game/LootLockerSDKManager.cs:1218]: This is copy pasta error, will fix :)

File: Runtime/Game/LootLockerSDKManager.cs [hunk 4]
```diff
@@ -1209,6 +1209,151 @@ public static void DeletePlayerFile(int fileId, Action<LootLockerResponse> onCom
...
+        /// <param name="onComplete">onComplete Action for handling the response of type LootLockerPaginatedPlayerProgressions</param>
+        public static void GetAllPlayerProgressions(int count, string after, Action<LootLockerPaginatedPlayerProgressions> onComplete)
```
[the remark this answers is not in the document]
Comment (kristijan-ujevic) [reply Runtime/Game/LootLockerSDKManager.cs:1220]: We are using ULID's as ID's on the backend

File: Runtime/Game/LootLockerSDKManager.cs [hunk 5]
```diff
@@ -1209,6 +1209,151 @@ public static void DeletePlayerFile(int fileId, Action<LootLockerResponse> onCom
...
+        {
+            if (!CheckInitialized())
+            {
+                onComplete?.Invoke(LootLockerResponseFactory.SDKNotInitializedError<LootLockerPaginatedPlayerProgressions>());
+                return;
+            }
+
+            var endpoint = LootLockerEndPoints.getAllPlayerProgressions.endPoint;
+
+            endpoint += "?";
```
[the remark this answers is not in the document]
Comment (kristijan-ujevic) [reply Runtime/Game/LootLockerSDKManager.cs:1230]: yes, tested and it works, this is something I'd like to discuss with you guys

File: Runtime/Game/LootLockerSDKManager.cs [hunk 6]
```diff
@@ -1209,6 +1209,151 @@ public static void DeletePlayerFile(int fileId, Action<LootLockerResponse> onCom
...
+            if (count > 0)
+                endpoint += $"count={count}&";
+
+            if (!string.IsNullOrEmpty(after))
+                endpoint += $"after={after}&";
```
[the remark this answers is not in the document]
Comment (kristijan-ujevic) [reply Runtime/Game/LootLockerSDKManager.cs:1235]: Same as above, we should discuss how to handle this, it works and the implementation is a lot cleaner by just adding the "&"

File: Runtime/Game/LootLockerSDKManager.cs [hunk 7]
```diff
@@ -1209,6 +1209,151 @@ public static void DeletePlayerFile(int fileId, Action<LootLockerResponse> onCom
...
+            var body = JsonConvert.SerializeObject(new { amount });  
+
+            LootLockerServerRequest.CallAPI(endpoint, LootLockerHTTPMethod.POST, body, onComplete: (serverResponse) => { LootLockerResponse.Serialize(onComplete, serverResponse); });
+        }
+        
+        /// <summary>
+        /// Subtracts points from a player progression.
+        /// </summary>
+        /// <param name="progressionKey">Progression key</param>
+        /// /// <param name="amount">Amount of points to be subtracted</param>
```
[the remark this answers is not in the document]
Comment (kristijan-ujevic) [reply Runtime/Game/LootLockerSDKManager.cs:1302]: copy pasta error :)

File: Runtime/Game/Requests/ProgressionsRequest.cs [hunk 8]
```diff
@@ -0,0 +1,161 @@
+using System;
+using System.Collections.Generic;
+using JetBrains.Annotations;
+
+namespace LootLocker.Requests
+{
+    public class LootLockerProgression : LootLockerResponse
+    {
+        public string id { get; set; }
```
[the remark this answers is not in the document]
Comment (kristijan-ujevic) [reply Runtime/Game/Requests/ProgressionsRequest.cs:9]: it is used for pagination

File: Runtime/Game/Requests/ProgressionsRequest.cs [hunk 9]
```diff
@@ -0,0 +1,161 @@
+using System;
+using System.Collections.Generic;
+using JetBrains.Annotations;
+
+namespace LootLocker.Requests
```
[the remark this answers is not in the document]
Comment (kristijan-ujevic) [reply Runtime/Game/Requests/ProgressionsRequest.cs:5]: Will do!

File: Runtime/Game/Requests/ProgressionsRequest.cs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        public string name { get; set; }
+        public bool active { get; set; }
+    }
+    
+    public class LootLockerPaginatedProgressions : LootLockerResponse
+    {
+        public LootLockerPagination pagination { get; set; }
+        public List<LootLockerProgression> items { get; set; }
+        
+        public class LootLockerProgression
```
[the remark this answers is not in the document]
Comment (kristijan-ujevic) [reply Runtime/Game/Requests/ProgressionsRequest.cs:20]: We should discuss this, I have a gripe LootLockerResponse too.

File: Runtime/Game/Requests/ProgressionsRequest.cs [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+        public class LootLockerProgressionTier
+        {
+            public string id { get; set; }
+            public int step { get; set; }
+            public int points_threshold { get; set; }
+            public LootLockerRewards rewards { get; set; }
+        }
+    }
+
+    public abstract class LLResponse
```
[the remark this answers is not in the document]
Comment (kristijan-ujevic) [reply Runtime/Game/Requests/ProgressionsRequest.cs:157]: Leftover from testing, will remove :D

**Review by kirre-bylund — APPROVED** [review]
Looking good Ma Dude!

**Review by kirre-bylund — APPROVED** [review]

**Comment by kirre-bylund** [comment]
Replaced by https://github.com/LootLocker/unity-sdk/pull/94

**Pull request closed without merging**